File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/r/test_02_multiple_cleanup_failures.R
# polyglot-family: errors_and_resources
# polyglot-concept: error_chaining_suppression_and_aggregation
# polyglot-related: languages/r/language/11_conditions_control_flow_and_cleanup/
# polyglot-related+: test_086_on_exit_multiple_cleanup_actions.R
#
# 共同问题：多个独立清理失败如何汇总而不丢失。
# 对照观察：base R 没有内建 ExceptionGroup；可逐项捕获 condition 并作为数据返回或再包装。

failures <- lapply(c("first", "second"), function(label) {
    tryCatch(stop(label), error = identity)
})

stopifnot(
    length(failures) == 2L,
    all(vapply(failures, inherits, logical(1), "error")),
    identical(vapply(failures, conditionMessage, ""), c("first", "second"))
)

File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/rust/test_02_multiple_cleanup_failures.rs
// polyglot-family: errors_and_resources
// polyglot-concept: error_chaining_suppression_and_aggregation
// polyglot-related: languages/rust/tests/language/08_errors_panics_and_resource_management/
// polyglot-related+: test_064_cleanup_failure_and_original_error.rs
//
// 共同问题：多个清理失败能否同时保留；遍历与匹配是否覆盖每个组成原因。
// 对照观察：std 没有通用 ExceptionGroup；应用类型可保存有序 error 集合并定义匹配策略。

#[derive(Debug, PartialEq)]
struct Aggregate(Vec<&'static str>);

#[test]
fn comparison() {
    let mut failures = Vec::new();
    for result in [Err::<(), _>("flush"), Err("close"), Ok(())] {
        if let Err(error) = result {
            failures.push(error);
        }
    }
    let aggregate = Aggregate(failures);
    assert_eq!(aggregate, Aggregate(vec!["flush", "close"]));
    assert!(aggregate.0.contains(&"close"));
}
